src/sysusers/sysusers.c

```diff
@@ -112,12 +112,12 @@ static int load_user_database(void) {
 
                 q = hashmap_put(database_uid, UID_TO_PTR(pw->pw_uid), n);
                 if (q < 0 && q != -EEXIST) {
-                        if (k < 0)
+                        if (k <= 0)
                                 free(n);
                         return q;
                 }
 
-                if (q < 0 && k < 0)
+                if (q <= 0 && k <= 0)
```

Comment:
`hashmap_remove` might help avoid calling `free`, but it also would make `systemd-sysusers` incorrect, so it seems that something else should be done.

Reply:
If `q <= 0 && k <= 0`, `n` wasn't added anywhere, and can be freed safely. This updated conditions appears correct.

Reply:
I think `q < 0 && k <= 0` is correct. `q == 0` seems to be something unexpected meaning that the address of the memory `n` points to has already been stored in `database_uid`. That memory is freed here and then in `free_database`, and also read somewhere in between.

Reply:
> If q <= 0 && k <= 0, n wasn't added anywhere, and can be freed safely

I'm not sure I get it. Could you expand on this? This particular check is supposed to make sure that the memory is freed if it's the second time a user or an id has been encountered.

Reply:
`q == 0 && k == 0` doesn't seem to be possible now just because there aren't any errors. It would be possible, though, if there were a bug somewhere. I believe `assert_se(q != 0 && k != 0)` would make it more clear that the condition isn't considered to be possible.

Reply:
`k == 0` is possible.

Reply:
That's true. Anyway, my point is that if something isn't considered to be possible and clearly indicates a programming error, it should probably be wrapped in `assert` or whatever else is used to detect programming errors.